src/features/weak-references.md

```diff
@@ -54,104 +54,133 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    const listener = (ev) => { this.events.push(ev); }
+    const weakRef = new WeakRef(listener);
+    const listenerWrapper = (ev) => {
+      const deref = weakRef.deref();
+      if (deref !== undefined) deref(ev);
+    });
+    socket.addEventListener("message", listenerWrapper);
   }
-});
+
+  // No more dispose!
+}
 ```
 
+We create a `WeakRef` whose _target_ is the actual event listener. Inside the `listenerWrapper`, we `deref` it. Because `WeakRef`s do not prevent garbage collection of their targets if the targets do not have other strong references, we must manually dereference them to get at the target. If the target has been garbage-collected in the meantime, `deref` returns `undefined`. Otherwise, the original target is returned, which is the `listener` function we then call.
+
+But there's still a problem here: we've added a level of indirection to `listener` by wrapping it a `WeakRef`, but `listenerWrapper` is still leaking for the same reason that `listener` was leaking originally. The solution to this is the companion feature to `WeakRef`, `FinalizationRegistry`. With the new `FinalizationRegistry` API, we can register a callback to run when the garbage collector zaps a register object. Such callbacks are known as _finalizers_.
+
 :::note
-**Note:** The `ref !== undefined && ref.deref() === undefined` is required because we could’ve added a new `WeakRef` with the same `name` between the old `WeakRef` enqueueing the finalization callback and actually running the finalization callback.
+**Note:** The finalization callback does not run immediately after garbage-collecting the event listener. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
 :::
 
-Our final implementation looks like this:
+We can register a callback with a `FinalizationRegistry` to remove `listenerWrapper` from the socket when the inner event listener is garbage-collected. Our final implementation looks like this:
 
 ```js
-const cache = new Map();
-
-const finalizationRegistry = new FinalizationRegistry((name) => {
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
+const gListenersRegistry = new FinalizationRegistry(
+  ({ socket, listenerWrapper }) => {
+    socket.removeEventListener(socket, listenerWrapper); // 5
+  });
+
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    const listener = (ev) => { this.events.push(ev); }
+    const weakRef = new WeakRef(listener); // 1
+    const listenerWrapper = (ev) => {
+      const deref = weakRef.deref(); // 2
+      if (deref !== undefined) deref(ev); // 3
+    });
+    socket.addEventListener("message", listenerWrapper);
+    gListenersRegistry.register(listener,
```

Comment:
Nitpick: such split, while common in C / C++, looks a bit odd in typical JS. I'd either put it on the same line at the risk of running over max-length

```js
gListenersRegistry.register(listener, { socket, listenerWrapper });
```

or use block-style formatting

```js
gListenersRegistry.register(listener, {
  socket,
  listenerWrapper
});
```

Reply:
Made one line.